### PR TITLE
Simplify sspicli interop marshaling

### PR DESCRIPTION
**Pull request by jkotas**
This makes it both faster and more AOT friendly

Contributes to https://github.com/dotnet/corert/issues/5674

### CODE REVIEW

File: src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs [hunk 1]
```diff
@@ -38,27 +36,23 @@ public static unsafe byte[] ToByteArray(List<SslApplicationProtocol> application
             }
 
             Sec_Application_Protocols protocols = new Sec_Application_Protocols();
-            protocols.ProtocolListsSize = (uint)(ProtocolListConstSize + protocolListSize);
-            protocols.ProtocolExtenstionType = ApplicationProtocolNegotiationExt.ALPN;
+            protocols.ProtocolListsSize = (uint)((sizeof(Sec_Application_Protocols) - sizeof(uint)) + protocolListSize);
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs]: The previous code was semi-self-documenting as to what was being calculated.  That's lost a bit here.  Can you add a comment?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCertificateHelper.cs [hunk 2]
```diff
@@ -55,11 +55,11 @@ internal static class WinHttpCertificateHelper
             unsafe
             {
                 var cppStruct = new Interop.Crypt32.CERT_CHAIN_POLICY_PARA();
-                cppStruct.cbSize = (uint)Marshal.SizeOf<Interop.Crypt32.CERT_CHAIN_POLICY_PARA>();
+                cppStruct.cbSize = (uint)sizeof(Interop.Crypt32.CERT_CHAIN_POLICY_PARA);
```
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCertificateHelper.cs]: What are the rules regarding when it is safe to use `sizeof()` vs. `Marshal.SizeOf<T>()` ?

File: src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs [hunk 3]
```diff
@@ -38,27 +36,23 @@ public static unsafe byte[] ToByteArray(List<SslApplicationProtocol> application
...
+            protocols.ProtocolExtensionType = ApplicationProtocolNegotiationExt.ALPN;
             protocols.ProtocolListSize = (short)protocolListSize;
 
-            Span<byte> pBuffer = new byte[protocolListSize];
+            byte[] buffer = new byte[sizeof(Sec_Application_Protocols) + protocolListSize];
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs]: This appears to be a larger value than it was before.  Was it broken before?

Comment (stephentoub) [reply src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs]: Oh, I see, this is replacing a second allocation made below. Nice.

Comment (jkotas) [reply src/Common/src/Interop/Windows/SChannel/Interop.Sec_Application_Protocols.cs]: The original code allocated two buffers: First temporary buffer was `protocolListSize`, and second final buffer was `sizeof(Sec_Application_Protocols) + protocolListSize`. I got rid of the temporary buffer allocation while I was on it.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCertificateHelper.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCertificateHelper.cs]: Basically you can use sizeof unless there is marshaling logic involved that would change the size of the type, e.g. if a field was attributed with MarshalAs that would convert it from one thing to another of a different size during the marshaling process.  If the type is blittable, sizeof is the way to go.

Comment (jkotas) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCertificateHelper.cs]: It is safe to use `sizeof` for [blittable types](https://en.wikipedia.org/wiki/Blittable_types). Both `sizeof()` and `Marshal.SizeOf<T>()` return same value for blittable types, except that sizeof is faster.

We have some of the best practices for interop perf documented in: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md#definition

**Comment by davidsh** [comment]
cc: @dotnet/ncl This is a good PR to learn about best-practice rules for optimum p/invokes.

**Comment by stephentoub** [comment]
@dotnet-bot test Windows x86 Release Build please
@dotnet-bot test OSX x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Comment by stephentoub** [comment]
Outerloop failures largely due to https://github.com/dotnet/corefx/issues/28631

**Pull request merged**